waffle_hub/hub/hub.py

```diff
@@ -1768,3 +1775,118 @@ def export_waffle(self, root_dir: str = None) -> ExportWaffleResult:
         result.waffle_file = self.waffle_file
 
         return result
+
+    def _hpo_hub_objective(self, **kwargs) -> float:
+        torch.cuda.empty_cache()
+
+        trial = kwargs.get("trial", None)
+        dataset = kwargs.get("dataset", None)
+        metric = kwargs.get("metric", None)
+        selected_params = kwargs.get("advance_params", None)
+
+        hub = Hub.new(
+            name=f"trial_{trial.number}",
+            root_dir=self.hub_dir / "hpo",
+            task=self.task,
+            model_type=self.model_type,
+            model_size=self.model_size,
+        )
+
+        train_result = hub.train(
+            dataset=dataset,
+            epochs=kwargs.get("epochs", None),
+            image_size=kwargs.get("image_size", None),
+            batch_size=kwargs.get("batch_size", None),
+            pretrained_model=kwargs.get("pretrained_model", None),
+            letter_box=kwargs.get("letter_box", False),
+            device=kwargs.get("device", "cpu"),
+            workers=kwargs.get("workers", 0),
+            advance_params=selected_params,
+            hold=kwargs.get("hold", True),
+        )
+
+        for metric_value in train_result.eval_metrics:
+            if metric_value["tag"] == metric:
+                result = metric_value["value"]
+                break
+        return float(result)
+
+    def hpo(
+        self,
+        dataset: Union[Dataset, str] = None,
+        dataset_root_dir: Union[str, Path] = None,
+        sampler: Union[dict, str] = None,
+        pruner: Union[dict, str] = None,
+        direction: str = None,
+        n_trials: int = None,
+        metric: str = None,
+        search_space: dict = None,
+        **kwargs,
+    ) -> dict:
+        """
+        Perform hyperparameter optimization (HPO) for the current task.
+
+        Args:
+            dataset (Union[Dataset, str], optional): The dataset to use for HPO. Can be a `Dataset` object or the name of a dataset. Defaults to None.
+            dataset_root_dir (Union[str, Path], optional): The root directory of the dataset. Defaults to None.
+            sampler (Union[dict, str], optional): The sampler to use for HPO. Can be a dictionary of sampler parameters or the name of a built-in sampler. Defaults to None.
+            pruner (Union[dict, str], optional): The pruner to use for HPO. Can be a dictionary of pruner parameters or the name of a built-in pruner. Defaults to None.
+            direction (str, optional): The direction of optimization. Can be 'maximize' or 'minimize'. Defaults to None.
+            n_trials (int, optional): The number of trials to run for HPO. Defaults to None.
+            metric (str, optional): The metric to optimize for. Defaults to None.
+            search_space (dict, optional): The search space for HPO. Defaults to None.
+            **kwargs: Additional keyword arguments to pass to the HPO function.
```

Comment:
What is the default value?

An error occurs when I insert the None value for each parameter.
I think you should keep `Optional` and set the Default value.

Reply:
Added default hpo config and modified to assign default value through setter.